### ISSUE TITLE
Increased memory usage on adding data

### ISSUE THREAD

**Issue by pooja1pathak** [ckan/ckan#5415]
I am using CKAN v2.7.2, installed by docker.

I have uploaded large data in dataset by GUI. On checking memory usage of ckan using ‘docker stats’ command, I have observed that the memory usage of ckan is increasing continuously and is not coming down.

I have following question:
1.	Is there any way to limit memory usage in ckan or by which we can decrease memory usage.
2.	Why Ckan memory usage is increased and never came down on addition of data.

**Comment by pooja1pathak** [comment]
I have another observation on this issue, when I execute get requests for dataset and groups constantly then the memory usage shown by ‘docker stats’ command also increase and does not come down to a lower value even after requests are completed.

Kindly provide the information related to this issue and the information related to questions asked earlier.

**Comment by kowh-ai** [comment]
Is this the same (or similar) as this issue: https://github.com/ckan/ckan/issues/5349 ?? Also, apart from the OS reporting more memory used by the CKAN processes, is there anything else that happens? ie: does the OS start swapping?

**Comment by pooja1pathak** [comment]
> Is this the same (or similar) as this issue: #5349 ?? 

Yes it is similar to the mentioned issue. But the actual cause was not known in that issue. Here, the cause of this memory increase is addition of large data. We are adding resource file of approx 9.75 MB constantly in production environment.

>Also, apart from the OS reporting more memory used by the CKAN processes, is there anything else that happens? ie: does the OS start swapping?

When all the memory is exhausted, CKAN container restarts as it is managed by Rancher. And after restarting, memory goes back to normal. But starts increasing again on addition of data.

**Comment by pooja1pathak** [comment]
Hi team, in our production environment we have a very large number of data and we constantly need to execute request to that data. As mentioned above, because of executing request on large data CKAN's memory consumption increases and when all the memory is exhausted then CKAN container restarts as it is managed by Rancher.

We've checked the number of process and child process they do not increase. So, we concluded that changing MaxConnectionsPerChild value will not help to limit memory usage.

Kindly guide us as in this to reduce or limit memory usage of CKAN.

**Comment by kowh-ai** [comment]
@pooja1pathak - I'm confused - Which Docker container are you changing "MaxConnectionsPerChild"? Are you using an additional Apache Web Server container somewhere? Are you following the 'Installing CKAN with Docker Compose' instructions?

**Comment by pooja1pathak** [comment]
> Which Docker container are you changing "MaxConnectionsPerChild"?

I am changing MaxConnectionsPerChild inside CKAN container.
>Are you using an additional Apache Web Server container somewhere?

No, we are not using any additional Apache Web Server Container

**Comment by kowh-ai** [comment]
@pooja1pathak Are you able to provide more detail? With performance problems like these it's very very difficult to be able to offer assistance as to where the problem may lie as it really could be anywhere. Are there any resource problems errors in the logs when the containers/system runs out of memory? Linux memory monitoring can be confusing at times as the OS tends to grab what it can when it needs it and it "looks like" it's not returning memory back but in fact it's just managed differently....Have you tried experimenting with any of the Performance Management solutions? AppDynamics is a very good one and there is a "free" version which may be able to hep you

**Comment by kowh-ai** [comment]
This may help you: https://www.linuxatemyram.com/

**Comment by pooja1pathak** [comment]
We have also observed memory increase in openknowledge/ckan-base:2.8.
 
We have tried using the memory profiler in source setup of ckan 2.7.2. We have observed that some functions in ckan increases memory and also they do not release memory. We have consolidated the data on basis of the memory increment. PFA of the same.
[memory_usage_by_files.xlsx](https://github.com/ckan/ckan/files/4995345/memory_usage_by_files.xlsx)

 
Kindly let us know if our observation is correct.

**Comment by pooja1pathak** [comment]
[get_request_with_apache.xlsx](https://github.com/ckan/ckan/files/5154803/get_request_with_apache.xlsx)
[get_request_without_apache.xlsx](https://github.com/ckan/ckan/files/5154804/get_request_without_apache.xlsx)
[GET_Request_Result.xlsx](https://github.com/ckan/ckan/files/5154806/GET_Request_Result.xlsx)


As communicated earlier, we are using the memory profiler in source setup of ckan 2.7.2. 

We have done setup with and without Apache and performed our investigation on both the setups.

We have collected data for 3 cycles of GET request on resource which we have uploaded in previous cycle. 
PFA of the files for data collected from profiler tool. We have consolidated the data based on memory usage in GET_Request_Result.xlsx.

As per our observation, "package_show" , “_resource_template”, “nav_link”, “check_access” and “format_resource_items” are not releasing memory.

Also, every time *current_url* method is called memory usage is increased. But neither "current_url" method nor any other method called before "current_url" is increasing any memory.

Kindly let us know if our observation is correct.
